src/Catalog/PackageCatalogItemCreator.cs

```diff
@@ -0,0 +1,234 @@
+            }
+
+            return new PackageCatalogItemCreator(httpClient, telemetryService, logger, storage);
+        }
+
+        public async Task<PackageCatalogItem> CreateAsync(
+            FeedPackageDetails packageItem,
+            DateTime timestamp,
+            CancellationToken cancellationToken)
+        {
+            if (packageItem == null)
+            {
+                throw new ArgumentNullException(nameof(packageItem));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            PackageCatalogItem item = null;
+
+            if (_storage != null)
+            {
+                item = await GetPackageViaStorageAsync(packageItem, cancellationToken);
+            }
+
+            if (item == null)
+            {
+                item = await GetPackageViaHttpAsync(packageItem, timestamp, item, cancellationToken);
+            }
+
+            return item;
+        }
+
+        private async Task<PackageCatalogItem> GetPackageViaStorageAsync(FeedPackageDetails packageItem, CancellationToken cancellationToken)
+        {
+            PackageCatalogItem item = null;
+            var packageName = $"{packageItem.PackageId.ToLowerInvariant()}.{packageItem.PackageVersion.ToLowerInvariant()}.nupkg";
+            var blob = _storage.GetCloudBlockBlobReference(packageName);
+
+            if (blob == null)
+            {
+                _telemetryService.TrackMetric(
+                    TelemetryConstants.NonExistentBlob,
+                    metric: 1,
+                    properties: GetProperties(packageItem, blob));
+            }
+            else
+            {
+                await blob.FetchAttributesAsync(cancellationToken);
+
+                string packageHash = null;
+                var etag = blob.ETag;
+
+                var metadata = await blob.GetMetadataAsync(cancellationToken);
+
+                if (metadata.TryGetValue("SHA512", out packageHash))
+                {
+                    using (var stream = await blob.GetStreamAsync(cancellationToken))
+                    {
+                        item = Utils.CreateCatalogItem(
+                            packageItem.ContentUri.ToString(),
+                            stream,
+                            packageItem.CreatedDate,
+                            packageItem.LastEditedDate,
+                            packageItem.PublishedDate,
+                            licenseNames: null,
+                            licenseReportUrl: null,
+                            packageHash: packageHash);
+
+                        if (item == null)
+                        {
+                            _logger.LogWarning("Unable to extract metadata from: {PackageDetailsContentUri}", packageItem.ContentUri);
+                        }
+                    }
+
+                    if (item != null)
+                    {
+                        await blob.FetchAttributesAsync(cancellationToken);
```

Comment:
Why are we fetching the blob again just for the purpose of checking its `ETag`?

Do we have any reason to believe that the `.nupkg` associated with a package will change as Feed2Catalog is processing it and the change to that `.nupkg` will not be reflected in a future commit to the catalog?

I would believe that if a package's `.nupkg` is modified while Feed2Catalog is processing it, one of these two scenarios has happened:
1. the package was modified by a revalidation (e.g. repo signing) that finished just as Feed2Catalog started processing the package
2. the package was deleted right when Feed2Catalog started processing it

Either cases will result in an additional catalog entry (because the packages container is modified BEFORE the feed is changed) so we don't need to check the blob again.

Reply:
1.  We are _not_ fetching the blob again just for the purpose of checking its ETag.  We are fetching the blob's _properties and metadata_ again for the purpose of checking its ETag.  There's a difference, which becomes more obvious with large blobs.  See [`CloudBlob.FetchAttributesAsync(...)`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.storage.blob.cloudblob.fetchattributesasync?view=azure-dotnet#Microsoft_WindowsAzure_Storage_Blob_CloudBlob_FetchAttributesAsync_System_Threading_CancellationToken_) for details.

2.  The original code fetched the package, wrote it locally, and performed multiple actions on the local file stream.  Implicitly, this provided a guarantee that the input for multiple subsequent actions was constant.  It's theoretically possible (though rare, I think) that in your first scenario, Feed2Catalog has read the package hash from blob v1.0 and a revalidation updates the blob while Feed2Catalog is reading it.  Now the package hash doesn't match the package, and the subsequent catalog entry will be incorrect.  This is more likely with large packages.  The catalog will be updated with correct information in time, but why break an existing consistency guarantee?

Reply:
Ok, so to clarify:

1. when you are reading a blob's stream, there's no guarantee that the contents of the blob are not modified while you are reading it
2. by fetching the blob's metadata again after the operation has completed, we guarantee that the stream we read data from was not modified while we were reading it

This makes sense, but you should definitely add a comment in the code to explain why we are fetching the metadata again.